service/resource/endpoint/create.go

```diff
@@ -1,11 +1,73 @@
 package endpoint
 
-import "context"
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiv1 "k8s.io/client-go/pkg/api/v1"
+)
 
 func (r *Resource) GetCreateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error) {
-	return nil, nil
+	currentEndpoints, err := toEndpointSlice(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredEndpoints, err := toEndpointSlice(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	createState := []Endpoint{}
+	for _, currentEndpoint := range currentEndpoints {
+		if !containsEndpoint(createState, currentEndpoint) {
+			createState = append(createState, currentEndpoint)
+		}
+	}
+	for _, desiredEndpoint := range desiredEndpoints {
+		if !containsEndpoint(createState, desiredEndpoint) {
+			createState = append(createState, desiredEndpoint)
+		}
+	}
```

Comment:
This looks conceptually wrong to me. I think create or update state should be the diff of current and desired state. In this case I'd expect createState to be desiredEndpoints that not exist in currentEndpoints. I guess I miss bigger picture there.

Reply:
Yes, the Problem was, that the desired state should be created using the resource which is watched (if I understood Tim correctly). The resource we watch is a pod with an annotation (Servicename and its IP), so the desiredState can only have the one entry without accessing the currentState. This is why I chose to uniquely concat both states instead of doing a diff. Dors that make sense to you?

Reply:
I understand. I agree it watches pods. But I would argue the actual state are endpoints not the pods. The resource doesn't change pods. I'd say the current state is the endpoint state, the desired state is the endpoint state. Diff is `desiredEnpoints - currentEndpoints` which should emit json patch which should be applied with kubernetes api in ProcessUpdateState.

Reply:
Okay either I misunderstood Tim / you here or there are two conflicting opinions on how to do this in `operatorkit`. Tim wrote the following:
> - The custom object is always a pod because the informer is set up accordingly.
> -   Using the current state (which I presume you want to do based on this code) to make up the desired state should never be done because the concept of current and desired state is fundamentally different.
> -  I feel like we are moving into the wrong direction. I think GetCurrentState should fetch the real endpoint with the real IP information. GetDesiredState should do what GetCurrentState is doing and construct the patch state using the IP that is potentially missing in the real endpoint by looking up the pod annotations. We would then compare the IPs in the other state creation methods and decide which IPs to forward to the processing methods. I apparently lead you into the wrong direction a bit.

This to me meant that I shouldn't fetch the `currentState` (which is the actual endpoint) inside the `desiredState`. Now you are suggesting that I should fetch the endpoint inside the `desiredState` in order to do a proper diff in the `getCreateState`. I could be completely misunderstanding either/both of you here.

Reply:
Can I just commit here and move forwards? I can create an issue to rework this in the future, but I'd like to get this done.

Reply:
I can interpret Tim's comment in several ways. So I won't be helpful there.

I dislike this is an update, but happens in ProcessCreate. We should fix that.

For me it'd be also fine to make GetCurrentState an noop. We don't make a diff of current and desired state, only desired state matters. My understanding of reconciliation in general is that you take current state, the desired state and make operations to transform current state into desired one. In this case current state doesn't really matter, as we change all IPs every time. Do you agree?

Reply:
I don't follow. In my current code every IP is changed every time as the result of `GetCreateState` contains all IPs. Then I completely replace all `EndpointAddresses` with those IPs. I don't see how your proposed change is any different other than putting the current state I have right now in the desired state as well.
 
What do you mean that this is an update and happens in `ProcessCreate`? AFAIK there is no way for me to just create the IPs here without fetching the `Endpoint`, so every action on the IPs would always be an update on the `Endpoint`. Or am I missing something?